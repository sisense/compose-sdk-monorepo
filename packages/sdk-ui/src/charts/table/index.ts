export * from './pure_table';
export type { TableProps } from './types';
export type { TableDesignOptions } from '../../chart-options-processor/translations/design_options';
