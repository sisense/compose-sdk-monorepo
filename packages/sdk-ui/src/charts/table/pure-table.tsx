import { useLayoutEffect, useMemo, useRef } from 'react';

import { TableDesignOptions } from '../../chart-options-processor/translations/design-options';
import { DataTableWrapper } from './data-table-wrapper';
import { formatNumbers } from './helpers/format-numbers';
import styles from './styles/table-chart.module.scss';
import { TableProps } from './types';

/**
 * PureTable Component. Table without any data fetch or aggregation logic.
 *
 * @returns Pure Table component
 */
export const PureTable = ({
  dataTable,
  dataOptions,
  designOptions = {} as TableDesignOptions,
  onSortUpdate,
  themeSettings,
  width = 400,
  height = 500,
}: TableProps) => {
  const listRef = useRef<HTMLDivElement>(null);
  const loadedTable = useRef(dataTable);

  const formattedTable = useMemo(
    () => formatNumbers(dataTable, dataOptions),
    [dataTable, dataOptions],
  );

  useLayoutEffect(() => {
    loadedTable.current = dataTable;
  }, [dataTable]);

  const customStyles = useMemo(
    () => ({
      sortIcon: 'caret' as const,
      showFieldTypeIcon: false,
      ...designOptions,
    }),
    [designOptions],
  );

  return (
    <div className={styles.component} ref={listRef}>
      <DataTableWrapper
        dataTable={formattedTable}
        dataOptions={dataOptions}
        onSortUpdate={onSortUpdate}
        height={height}
        width={width}
        customStyles={customStyles}
        themeSettings={themeSettings}
      />
    </div>
  );
};
