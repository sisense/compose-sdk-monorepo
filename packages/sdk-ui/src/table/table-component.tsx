import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import Pagination from '@mui/material/Pagination';
import { isDataSource } from '@ethings-os/sdk-data';
import { TableProps } from '../props';
import { useThemeContext } from '../theme-provider';
import { translateTableDataOptions } from '../chart-data-options/translate-data-options';
import { useTableData } from './hooks/use-table-data';
import { useTableDataTable } from './hooks/use-table-datatable';
import { Column as DataTableColumn } from '../chart-data-processor/table-processor';
import { updateInnerDataOptionsSort } from '../chart-data/table-data';
import { PureTable } from '../charts/table';
import { getCustomPaginationStyles } from './styles/get-custom-pagination-styles';
import { generateUniqueDataColumnsNames } from '../chart-data-options/validate-data-options';
import { StyledMeasureColumn, TableDataOptionsInternal } from '../chart-data-options/types';
import { isDataTableEmpty } from '../chart-data-processor/table-creators';
import { NoResultsOverlay } from '../no-results-overlay/no-results-overlay';
import { DynamicSizeContainer, getChartDefaultSize } from '../dynamic-size-container';
import { LoadingIndicator } from '../common/components/loading-indicator';
import { getFilterListAndRelationsJaql } from '@ethings-os/sdk-data';
import { translateTableStyleOptionsToDesignOptions } from './translations/design-options';
import { orderBy } from '../chart-data-processor/table-processor';
import { isMeasureColumn, translateColumnToMeasure } from '@/chart-data-options/utils';
import { TranslatableError } from '@/translation/translatable-error';
import { isData } from '@/chart/regular-chart';

export const DEFAULT_TABLE_ROWS_PER_PAGE = 25;

/** How many pages of data will be loaded in one query */
export const PAGES_BATCH_SIZE = 10;

/**
 * Component that renders a table with aggregation and pagination.
 */
export const TableComponent = ({
  dataSet,
  dataOptions,
  styleOptions = {},
  filters,
  onDataReady,
}: TableProps) => {
  const { rowsPerPage = DEFAULT_TABLE_ROWS_PER_PAGE, width, height } = styleOptions;
  const { themeSettings } = useThemeContext();
  const [offset, setOffset] = useState(0);
  const { filters: filterList, relations: filterRelations } =
    getFilterListAndRelationsJaql(filters);
  const [currentPage, setCurrentPage] = useState(1);
  const paginationEl = useRef(null);

  const translatedDataOptions = useMemo(
    () => translateTableDataOptions(dataOptions),
    [dataOptions],
  );
  const dataColumnNamesMapping = useMemo(
    () =>
      generateUniqueDataColumnsNames(
        (translatedDataOptions.columns.filter(isMeasureColumn) as StyledMeasureColumn[]).map(
          translateColumnToMeasure,
        ),
      ),
    [translatedDataOptions],
  );

  const designOptions = useMemo(
    () => translateTableStyleOptionsToDesignOptions(styleOptions),
    [styleOptions],
  );

  const [innerDataOptions, setInnerDataOptions] =
    useState<TableDataOptionsInternal>(translatedDataOptions);

  const [usedDataSet, setUsedDataset] = useState(dataSet);
  const [data, updatedDataOptions] = useTableData({
    dataSet: usedDataSet,
    dataOptions: innerDataOptions,
    filters: filterList,
    filterRelations,
    count: rowsPerPage * PAGES_BATCH_SIZE,
    offset,
  });

  const finalData = useMemo(() => {
    if (data && onDataReady) {
      const customizedData = onDataReady(data);
      if (!isData(customizedData)) {
        throw new TranslatableError('errors.incorrectOnDataReadyHandler');
      }
      return customizedData;
    }
    return data;
    // Ignore rule to avoid unnecessary calls "onDataReady"
    // Trigger only on "data" update
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data]);

  const dataTable = useTableDataTable({
    data: finalData,
    innerDataOptions: updatedDataOptions,
    dataColumnNamesMapping,
    needToAggregate: !isDataSource(usedDataSet),
  });

  useEffect(() => {
    setInnerDataOptions(translatedDataOptions);
    setUsedDataset(dataSet);

    setCurrentPage(1);
    setOffset(0);
  }, [dataSet, translatedDataOptions]);

  const onPageChange = useCallback(
    (page: number) => {
      if (!data) return;
      const lastPage = Math.floor(data.rows.length / rowsPerPage) === page;
      if (lastPage) {
        setOffset(data.rows.length);
      }
      setCurrentPage(page);
    },
    [data, rowsPerPage],
  );

  const onSortUpdate = useCallback(
    (column: DataTableColumn) => {
      setCurrentPage(1);
      setOffset(0);
      if (innerDataOptions) {
        setInnerDataOptions(updateInnerDataOptionsSort(innerDataOptions, column));
      }
    },
    [innerDataOptions],
  );

  const sortedTable = useMemo(() => {
    if (!dataTable) {
      return dataTable;
    }
    if (isDataSource(usedDataSet)) {
      return dataTable;
    }
    const columnWithSorting = dataTable.columns.find((c) => c.direction !== 0);
    if (!columnWithSorting) {
      return dataTable;
    }
    return orderBy(dataTable, [columnWithSorting]);
  }, [dataTable, usedDataSet]);

  const paginatedTable = useMemo(
    () =>
      sortedTable
        ? {
            columns: sortedTable.columns,
            rows: sortedTable.rows.slice(
              (currentPage - 1) * rowsPerPage,
              currentPage * rowsPerPage,
            ),
          }
        : undefined,
    [sortedTable, currentPage, rowsPerPage],
  );

  if (!updatedDataOptions) return null;

  return (
    <DynamicSizeContainer
      defaultSize={getChartDefaultSize('table')}
      size={{
        width,
        height,
      }}
    >
      {(size) => {
        if (!dataTable || !paginatedTable) {
          return <LoadingIndicator themeSettings={themeSettings} />;
        }

        if (isDataTableEmpty(dataTable)) {
          return <NoResultsOverlay iconType={'table'} />;
        }

        const pagesCount = Math.ceil(dataTable.rows.length / rowsPerPage);
        const paginationHeight = 32;

        return (
          <div
            style={{
              display: 'flex',
              flexDirection: 'column',
              alignItems: 'center',
              backgroundColor: themeSettings.chart.backgroundColor,
            }}
            aria-label="table-root"
          >
            <PureTable
              dataTable={paginatedTable}
              dataOptions={updatedDataOptions}
              designOptions={designOptions}
              themeSettings={themeSettings}
              width={size.width}
              height={size.height - paginationHeight}
              onSortUpdate={onSortUpdate}
            />
            <Pagination
              ref={paginationEl}
              page={currentPage}
              count={pagesCount}
              onChange={(event, page) => onPageChange(page)}
              sx={getCustomPaginationStyles(themeSettings)}
            />
          </div>
        );
      }}
    </DynamicSizeContainer>
  );
};
