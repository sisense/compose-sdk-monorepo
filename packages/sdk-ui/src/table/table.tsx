/* eslint-disable max-lines-per-function */
import { Data } from '@sisense/sdk-data';

import { asSisenseComponent } from '../decorators/component-decorators/as-sisense-component';
import { TableProps } from '../props';
import { TableComponent } from './table-component';

/** Function to check if we should wait for sisense context for rendering the table */
function shouldSkipSisenseContextWaiting(props: TableProps) {
  return isCompleteDataSet(props.dataSet);
}

function isCompleteDataSet(dataSet: TableProps['dataSet']): dataSet is Data {
  return !!dataSet && typeof dataSet !== 'string' && 'rows' in dataSet && 'columns' in dataSet;
}

/**
 * Table with aggregation and pagination.
 *
 * ## Example
 *
 * Table displaying year, condition, and total revenue from the Sample ECommerce data model.
 *
 * <iframe
 *  src='https://csdk-playground.sisense.com/?example=tables%2Faggregated-table&mode=docs'
 *  width=800
 *  height=870
 *  style='border:none;'
 * />
 *
 * @param props - Table properties
 * @returns Table component
 * @group Data Grids
 */

export const Table = asSisenseComponent({
  componentName: 'Table',
  shouldSkipSisenseContextWaiting,
})((props: TableProps) => {
  return <TableComponent {...props} />;
});

export const DEFAULT_TABLE_ROWS_PER_PAGE = 25 as const;

/** How many pages of data will be loaded in one query */
export const PAGES_BATCH_SIZE = 10;
