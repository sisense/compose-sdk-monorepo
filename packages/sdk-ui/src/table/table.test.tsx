/** @vitest-environment jsdom */
import { render } from '@testing-library/react';

import { Table } from './table';

// Table tests
describe('Table', () => {
  const col1 = { name: 'AgeRange', type: 'string' };
  const col2 = { name: 'Cost', type: 'number' };
  const dataSet = {
    columns: [col1, col2],
    rows: [
      ['0-18', 1000],
      ['19-28', 19.123],
      ['29-35', 125],
    ],
  };

  it('should render Table', async () => {
    const tableDataOptions = {
      columns: [col1, col2],
    };
    const { findByLabelText, findAllByRole } = render(
      <Table dataSet={dataSet} dataOptions={tableDataOptions} />,
    );
    const table = await findByLabelText('table-root');
    expect(table).toBeTruthy();
    const columns = await findAllByRole('columnheader');
    expect(columns.length).toBe(tableDataOptions.columns.length);
    const rows = await findAllByRole('row');
    expect(rows.length).toBe(dataSet.rows.length + 1); // +1 for header row
  });

  it('should render Table with base props', async () => {
    const table = render(
      <Table
        dataSet={dataSet}
        dataOptions={{
          columns: [col1],
        }}
      />,
    );

    const heading = await table.findByText('AgeRange');
    expect(heading).toBeTruthy();

    dataSet.rows.forEach(([value]) => {
      const cell = table.queryByText(value);
      expect(cell).toBeTruthy();
    });
  });

  it('should render with error when provided column missing in data', async () => {
    const spy = vi.spyOn(console, 'error');
    let errorThrown = false;
    spy.mockImplementation(() => {
      errorThrown = true;
    });

    const { container, findByLabelText } = render(
      <Table
        dataSet={dataSet}
        dataOptions={{
          columns: [{ name: 'NotExist', type: 'string' }],
        }}
      />,
    );
    const errorBox = await findByLabelText('error-box');
    expect(errorBox).toBeTruthy();
    const tableWrapper = container.querySelector('.tableWrapper');
    expect(tableWrapper).toBeFalsy();
    expect(errorThrown).toBeTruthy();

    spy.mockRestore();
  });

  it('should show No Results overlay in Table when data missing', async () => {
    const { findByText } = render(
      <Table
        dataSet={{
          columns: dataSet.columns,
          rows: [],
        }}
        dataOptions={{
          columns: [col1],
        }}
      />,
    );
    const overlayTitle = await findByText('No Results');
    expect(overlayTitle).toBeTruthy();
  });

  it('should correctly react on data options change', async () => {
    const { findByText, rerender } = render(
      <Table
        dataSet={{
          ...dataSet,
          columns: [col1],
        }}
        dataOptions={{ columns: [col1] }}
      />,
    );

    const headerFromFirstRender = await findByText('AgeRange');
    expect(headerFromFirstRender).toBeTruthy();

    rerender(
      <Table
        dataSet={{
          ...dataSet,
          columns: [col2],
        }}
        dataOptions={{ columns: [col2] }}
      />,
    );

    const headerFromSecondRender = await findByText('Cost');
    expect(headerFromSecondRender).toBeTruthy();
  });

  it('should trigger onDataReady callback and render Table with modified data', async () => {
    const MODIFIED_VALUE = 'Modified Table Value';
    const onDataReadyMock = vi.fn().mockImplementation((data) => {
      expect(data).toEqual(dataSet);
      return {
        ...data,
        rows: [[MODIFIED_VALUE, 1]],
      };
    });

    const { findByText } = render(
      <Table
        dataSet={dataSet}
        dataOptions={{ columns: [col1, col2] }}
        onDataReady={onDataReadyMock}
      />,
    );

    const renderedModifiedValue = await findByText(MODIFIED_VALUE);
    expect(renderedModifiedValue).toBeTruthy();
    expect(onDataReadyMock).toHaveBeenCalledTimes(1);
  });

  it('should trigger onDataReady callback and render error for incorrect callback return', async () => {
    const onDataReadyMock = vi.fn().mockImplementation(() => undefined);

    const { findByLabelText } = render(
      <Table
        dataSet={dataSet}
        dataOptions={{ columns: [col1, col2] }}
        onDataReady={onDataReadyMock}
      />,
    );

    const errorBoxContainer = await findByLabelText('error-box');
    expect(errorBoxContainer).toBeTruthy();
    expect(onDataReadyMock).toHaveBeenCalledTimes(2);
  });
});
