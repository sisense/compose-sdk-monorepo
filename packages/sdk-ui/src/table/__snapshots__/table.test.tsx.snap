// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > should render Table 1`] = `
<div>
  <div
    style="width: 100%; height: 100%;"
  >
    <div
      style="width: 400px; height: 500px; position: relative;"
    >
      <div
        style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
      >
        <div
          aria-label="table-root"
          style="display: flex; flex-direction: column; align-items: center; background-color: rgb(255, 255, 255);"
        >
          <div
            class="_component_2a48df"
          >
            <div
              class="_tableWrapper_4b88bf"
              style="padding: 20px 20px;"
            >
              <div
                aria-rowcount="4"
                class="_table_4b88bf fixedDataTableLayout_main public_fixedDataTable_main"
                role="grid"
                style="height: 460px; width: 360px;"
              >
                <div
                  class="fixedDataTableLayout_rowsContainer"
                  style="height: 458px; width: 360px;"
                >
                  <div
                    class="fixedDataTableColumnResizerLineLayout_main fixedDataTableColumnResizerLineLayout_hiddenElem public_fixedDataTableColumnResizerLine_main"
                    style="width: 0px; height: 460px; left: 0px;"
                  >
                    <div
                      class="fixedDataTableColumnResizerLineLayout_mouseArea"
                      style="height: 460px;"
                    />
                  </div>
                  <div
                    class="fixedDataTableRowLayout_rowWrapper"
                    style="width: 360px; height: 26px; z-index: 1; visibility: visible; transform: translate3d(0px,0px,0);"
                  >
                    <div
                      aria-rowindex="1"
                      class="fixedDataTableRowLayout_main public_fixedDataTableRow_main fixedDataTableLayout_header public_fixedDataTable_header"
                      role="row"
                      style="width: 360px; height: 26px;"
                    >
                      <div
                        class="fixedDataTableRowLayout_body"
                      >
                        <div
                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                          style="height: 26px; width: 0px; left: 0px;"
                        >
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroup"
                            style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                          />
                        </div>
                        <div
                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                          style="height: 26px; width: 360px; left: 0px;"
                        >
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroup"
                            style="height: 26px; position: absolute; width: 252px; z-index: 0; transform: translate3d(0px,0px,0);"
                          >
                            <div
                              class="fixedDataTableCellLayout_main public_fixedDataTableCell_main"
                              role="columnheader"
                              style="height: 26px; width: 126px; left: 0px;"
                            >
                              <div
                                class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableHeader_4b88bf"
                                style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                              >
                                <div
                                  class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                >
                                  <div
                                    class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                  >
                                    <div
                                      class="public_fixedDataTableCell_cellContent"
                                    >
                                      <div
                                        class="_headerCell_d9b667"
                                      >
                                        <div
                                          class="_headerLabel_d9b667"
                                        >
                                          <div
                                            class="_tableCellContent_4b88bf"
                                            style="max-width: 90px;"
                                          >
                                            <div>
                                              AgeRange
                                            </div>
                                          </div>
                                        </div>
                                        <div
                                          class="_sortIcon_d9b667 _caret_d9b667 _noSort_d9b667"
                                        />
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                            <div
                              class="fixedDataTableCellLayout_main fixedDataTableCellLayout_alignRight public_fixedDataTableCell_alignRight public_fixedDataTableCell_main"
                              role="columnheader"
                              style="height: 26px; width: 126px; left: 126px;"
                            >
                              <div
                                class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableHeader_4b88bf"
                                style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                              >
                                <div
                                  class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                >
                                  <div
                                    class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                  >
                                    <div
                                      class="public_fixedDataTableCell_cellContent"
                                    >
                                      <div
                                        class="_headerCell_d9b667"
                                      >
                                        <div
                                          class="_headerLabel_d9b667"
                                        >
                                          <div
                                            class="_tableCellContent_4b88bf"
                                            style="max-width: 90px;"
                                          >
                                            <div>
                                              Cost
                                            </div>
                                          </div>
                                        </div>
                                        <div
                                          class="_sortIcon_d9b667 _caret_d9b667 _noSort_d9b667"
                                        />
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class=""
                          style="left: 0px; height: 26px;"
                        />
                        <div
                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                          style="height: 26px; width: 0px; left: 360px;"
                        >
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroup"
                            style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    style="transform: translate3d(0px,26px,0);"
                  >
                    <div
                      class="fixedDataTableRowLayout_rowWrapper"
                      style="width: 360px; height: 26px; z-index: 0; visibility: visible; transform: translate3d(0px,0px,0);"
                    >
                      <div
                        aria-rowindex="2"
                        class="fixedDataTableRowLayout_main public_fixedDataTableRow_main public_fixedDataTableRow_even public_fixedDataTable_bodyRow"
                        role="row"
                        style="width: 360px; height: 26px;"
                      >
                        <div
                          class="fixedDataTableRowLayout_body"
                        >
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 0px; left: 0px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                            />
                          </div>
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 360px; left: 0px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 252px; z-index: 0; transform: translate3d(0px,0px,0);"
                            >
                              <div
                                class="fixedDataTableCellLayout_main public_fixedDataTableCell_main"
                                role="gridcell"
                                style="height: 26px; width: 126px; left: 0px;"
                              >
                                <div
                                  class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                  style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                >
                                  <div
                                    class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                  >
                                    <div
                                      class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                    >
                                      <div
                                        class="public_fixedDataTableCell_cellContent"
                                      >
                                        <div
                                          class="_tableCellContent_4b88bf"
                                          style="max-width: 102px;"
                                        >
                                          <div>
                                            0-18
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                              <div
                                class="fixedDataTableCellLayout_main fixedDataTableCellLayout_alignRight public_fixedDataTableCell_alignRight public_fixedDataTableCell_main"
                                role="gridcell"
                                style="height: 26px; width: 126px; left: 126px;"
                              >
                                <div
                                  class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                  style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                >
                                  <div
                                    class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                  >
                                    <div
                                      class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                    >
                                      <div
                                        class="public_fixedDataTableCell_cellContent"
                                      >
                                        <div
                                          class="_tableCellContent_4b88bf"
                                          style="max-width: 102px;"
                                        >
                                          <div>
                                            1K
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class=""
                            style="left: 0px; height: 26px;"
                          />
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 0px; left: 360px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="fixedDataTableRowLayout_rowWrapper"
                      style="width: 360px; height: 26px; z-index: 0; visibility: visible; transform: translate3d(0px,26px,0);"
                    >
                      <div
                        aria-rowindex="3"
                        class="fixedDataTableRowLayout_main public_fixedDataTableRow_main public_fixedDataTableRow_highlighted public_fixedDataTableRow_odd public_fixedDataTable_bodyRow"
                        role="row"
                        style="width: 360px; height: 26px;"
                      >
                        <div
                          class="fixedDataTableRowLayout_body"
                        >
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 0px; left: 0px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                            />
                          </div>
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 360px; left: 0px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 252px; z-index: 0; transform: translate3d(0px,0px,0);"
                            >
                              <div
                                class="fixedDataTableCellLayout_main public_fixedDataTableCell_main"
                                role="gridcell"
                                style="height: 26px; width: 126px; left: 0px;"
                              >
                                <div
                                  class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                  style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                >
                                  <div
                                    class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                  >
                                    <div
                                      class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                    >
                                      <div
                                        class="public_fixedDataTableCell_cellContent"
                                      >
                                        <div
                                          class="_tableCellContent_4b88bf"
                                          style="max-width: 102px;"
                                        >
                                          <div>
                                            19-28
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                              <div
                                class="fixedDataTableCellLayout_main fixedDataTableCellLayout_alignRight public_fixedDataTableCell_alignRight public_fixedDataTableCell_main"
                                role="gridcell"
                                style="height: 26px; width: 126px; left: 126px;"
                              >
                                <div
                                  class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                  style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                >
                                  <div
                                    class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                  >
                                    <div
                                      class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                    >
                                      <div
                                        class="public_fixedDataTableCell_cellContent"
                                      >
                                        <div
                                          class="_tableCellContent_4b88bf"
                                          style="max-width: 102px;"
                                        >
                                          <div>
                                            19.12
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class=""
                            style="left: 0px; height: 26px;"
                          />
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 0px; left: 360px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="fixedDataTableRowLayout_rowWrapper"
                      style="width: 360px; height: 26px; z-index: 0; visibility: visible; transform: translate3d(0px,52px,0);"
                    >
                      <div
                        aria-rowindex="4"
                        class="fixedDataTableRowLayout_main public_fixedDataTableRow_main public_fixedDataTableRow_even public_fixedDataTable_bodyRow fixedDataTableLayout_hasBottomBorder public_fixedDataTable_hasBottomBorder"
                        role="row"
                        style="width: 360px; height: 26px;"
                      >
                        <div
                          class="fixedDataTableRowLayout_body"
                        >
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 0px; left: 0px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                            />
                          </div>
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 360px; left: 0px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 252px; z-index: 0; transform: translate3d(0px,0px,0);"
                            >
                              <div
                                class="fixedDataTableCellLayout_main public_fixedDataTableCell_main"
                                role="gridcell"
                                style="height: 26px; width: 126px; left: 0px;"
                              >
                                <div
                                  class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                  style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                >
                                  <div
                                    class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                  >
                                    <div
                                      class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                    >
                                      <div
                                        class="public_fixedDataTableCell_cellContent"
                                      >
                                        <div
                                          class="_tableCellContent_4b88bf"
                                          style="max-width: 102px;"
                                        >
                                          <div>
                                            29-35
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                              <div
                                class="fixedDataTableCellLayout_main fixedDataTableCellLayout_alignRight public_fixedDataTableCell_alignRight public_fixedDataTableCell_main"
                                role="gridcell"
                                style="height: 26px; width: 126px; left: 126px;"
                              >
                                <div
                                  class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                  style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                >
                                  <div
                                    class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                  >
                                    <div
                                      class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                    >
                                      <div
                                        class="public_fixedDataTableCell_cellContent"
                                      >
                                        <div
                                          class="_tableCellContent_4b88bf"
                                          style="max-width: 102px;"
                                        >
                                          <div>
                                            125
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class=""
                            style="left: 0px; height: 26px;"
                          />
                          <div
                            class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                            style="height: 26px; width: 0px; left: 360px;"
                          >
                            <div
                              class="fixedDataTableCellGroupLayout_cellGroup"
                              style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
